pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh

```diff
@@ -0,0 +1,80 @@
+npmSetupHook() {
+    echo "Executing npmSetupHook"
+
+    runHook preConfigure
+
+    echo "Configuring npm"
+
+    export HOME=$TMPDIR
+    export npm_config_nodedir="@nodeSrc@"
+
+    npm config set cache "$npmDeps"
+    npm config set offline true
+    npm config set progress false
+
+    cacheLockfile="$npmDeps/package-lock.json"
+    srcLockfile="$NIX_BUILD_TOP/$sourceRoot/package-lock.json"
+
+    echo "Validating consistency between $srcLockfile and $cacheLockfile"
+
+    if ! @diff@ $srcLockfile $cacheLockfile; then
+      # If the diff failed, first double-check that the file exists, so we can
+      # give a friendlier error msg.
+      if ! [ -e $srcLockfile ]; then
+        echo "ERROR: Missing package-lock.json from src. Expected to find it at: $srcLockfile"
+        echo "Hint: You can use the patches attribute to add a package-lock.json manually to the build."
+        exit 1
+      fi
+
+      if ! [ -e $cacheLockfile ]; then
+        echo "ERROR: Missing lockfile from cache. Expected to find it at: $cacheLockfile"
+        exit 1
+      fi
+
+      echo
+      echo "ERROR: npmDepsHash is out of date"
+      echo
+      echo "package-lock.json is not the same in $npmDeps"
+      echo
+      echo "To fix the issue:"
+      echo '1. Use `lib.fakeHash` as the npmDepsHash value'
+      echo "2. Build the derivation and wait for it to fail with a hash mismatch"
+      echo "3. Copy the 'got: sha256-' value back into the npmDepsHash field"
+      echo
+
+      exit 1
+    fi
+
+    echo "Installing dependencies"
+
+    # Frozen install...
+    npm ci --ignore-scripts
+
+    # ...patch shebangs...
+    patchShebangs node_modules
+
+    root=$(pwd)
+
+    while IFS= read -r f; do
+        cd "$f" || continue
+
+        echo "$f"
+
+        scripts=("preinstall" "install" "postinstall")
+
+        for script in "${scripts[@]}"; do
+            if grep -q "\"$script\"" package.json; then
+                echo "  running $script script"
+                npm run "$script" --if-present
+            fi
+        done
+
+        cd "$root" || exit 1
+    done < <(python3 @getDepsScript@)
+
+    runHook postConfigure
+
+    echo "Finished npmSetupHook"
+}
+
+configurePhase=npmSetupHook
```

Comment:
This is the only phase that that wasn't added -- I'm not sure when you'd want to override this (and duplicate the effort in the phase), since it sets up everything, but for consistency I guess that makes sense.

Reply:
It might be worth adding corresponding `dont` options (or options that act as that) to all these phases as well (which I only currently do in two of them).

Though, now that I think about it, having both conditionals creates confusion:

1. When using `buildNpmPackage`, setting a custom `configurePhase` will cause the npm phase not to be used
2. When using this hook on its own, why would you set `dontNpmConfigure` instead of just... not including the hook?

I don't know, I'm confused. Cargo does this in its build hook, and I don't really see why both are needed.

@jonringer What do you think about this, as the person who added these options in #114716? It seems like these options don't really provide much benefit when hooks are used properly -- maybe I'm missing something?

Reply:
> It might be worth adding corresponding dont options (or options that act as that) to all these phases as well (which I only currently do in two of them).

I definitely agree, that's the most consistent with other builders as well AFAIK. Can't hurt to have a mechanism to disable these when using `buildNpmPackage` :)

> When using buildNpmPackage, setting a custom configurePhase will cause the npm phase not to be used

Yes, which might be preferable in some scenarios. That way you can have complete control over the build process, you can override the phase with custom behaviour that produces a similar result yourself.

> When using this hook on its own, why would you set dontNpmConfigure instead of just... not including the hook?

That does seem useless yeah, I can't imagine any scenario you'd do that.

Reply:
> Can't hurt to have a mechanism to disable these when using `buildNpmPackage`

To be clear, what I mean is that all of these should be disabled if a custom phase is already set -- having `dont` options is useless as long as those conditionals exist.